### ISSUE TITLE
bug: using root logger with debug level causes infinite hang

### ISSUE THREAD

**Issue by georgedorn** [kislyuk/watchtower#93]
Very similar to https://github.com/kislyuk/watchtower/issues/65 except that it happens during shutdown instead of init.

Also, because it is waiting on a queue to clear, it never hits a recursion limit and instead hangs forever.

More:

I'm using Django, and this triggers when leaving a shell.  By adding a file logger, I'm able to get this:

```
$ ./manage.py shell
Python 3.7.6 (default, Jan  8 2020, 19:59:22)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]:                                                                                                              
Do you really want to exit ([y]/n)?
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 2033, in shutdown
    h.flush()
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/watchtower/__init__.py", line 250, in flush
    q.join()
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/queue.py", line 89, in join
    self.all_tasks_done.wait()
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/threading.py", line 296, in wait
    waiter.acquire()
KeyboardInterrupt
```

And the contents of the log file:

```

2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event before-parameter-build.cloudwatch-logs.CreateLogStream: calling handler <function generate_idempotent_uuid at 0x7eff1de77320>
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event before-call.cloudwatch-logs.CreateLogStream: calling handler <function inject_api_version_header_if_needed at 0x7eff1de7bdd0>
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/endpoint.py:101] Making request for OperationModel(name=CreateLogStream) with params: {'url_path': '/', 'query_string': '', 'method': 'POST', 'headers': {'X-Amz-Target': 'Logs_20140328.CreateLogStream', 'Content-Type': 'application/x-amz-json-1.1', 'User-Agent': 'Boto3/1.12.35 Python/3.7.6 Linux/4.15.0-1058-aws Botocore/1.15.35'}, 'body': b'{"logGroupName": "dev", "logStreamName": "django"}', 'url': 'https://logs.us-west-2.amazonaws.com/', 'context': {'client_region': 'us-west-2', 'client_config': <botocore.config.Config object at 0x7eff1d108f10>, 'has_streaming_input': False, 'auth_type': None}}
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event request-created.cloudwatch-logs.CreateLogStream: calling handler <bound method RequestSigner.handler of <botocore.signers.RequestSigner object at 0x7eff1d108ed0>>
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event choose-signer.cloudwatch-logs.CreateLogStream: calling handler <function set_operation_specific_signer at 0x7eff1de77200>
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:364] Calculating signature using v4 auth.
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:365] CanonicalRequest:
POST
/

content-type:application/x-amz-json-1.1
host:logs.us-west-2.amazonaws.com
x-amz-date:20200402T234459Z
x-amz-security-token:[snip]==
x-amz-target:Logs_20140328.CreateLogStream

content-type;host;x-amz-date;x-amz-security-token;x-amz-target
[snip]
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:367] StringToSign:
AWS4-HMAC-SHA256
20200402T234459Z
20200402/us-west-2/logs/aws4_request
[snip]
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:369] Signature:
[snip
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/endpoint.py:187] Sending http request: <AWSPreparedRequest stream_output=False, method=POST, url=https://logs.us-west-2.amazonaws.com/, headers={'X-Amz-Target': b'Logs_20140328.CreateLogStream', 'Content-Type': b'application/x-amz-json-1.1', 'User-Agent': b'Boto3/1.12.35 Python/3.7.6 Linux/4.15.0-1058-aws Botocore/1.15.35', 'X-Amz-Date': b'20200402T234459Z', 'X-Amz-Security-Token': b'[snip]==', 'Authorization': b'AWS4-HMAC-SHA256 Credential=[snip]/20200402/us-west-2/logs/aws4_request, SignedHeaders=content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=[snip]', 'Content-Length': '50'}>
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/urllib3/connectionpool.py:266] Resetting dropped connection: logs.us-west-2.amazonaws.com
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/urllib3/connectionpool.py:437] https://logs.us-west-2.amazonaws.com:443 "POST / HTTP/1.1" 400 95
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/parsers.py:234] Response headers: {'x-amzn-RequestId': '[snip]', 'Content-Type': 'application/x-amz-json-1.1', 'Content-Length': '95', 'Date': 'Thu, 02 Apr 2020 23:43:16 GMT', 'Connection': 'close'}
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/parsers.py:235] Response body:
b'{"__type":"ResourceAlreadyExistsException","message":"The specified log stream already exists"}'
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event needs-retry.cloudwatch-logs.CreateLogStream: calling handler <botocore.retryhandler.RetryHandler object at 0x7eff1d136310>
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/retryhandler.py:187] No retry needed.
2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/asyncio/selector_events.py:58] Using selector: EpollSelector
2020-04-02 23:45:00 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/asyncio/selector_events.py:58] Using selector: EpollSelector
2020-04-02 23:45:01 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event before-parameter-build.cloudwatch-logs.PutLogEvents: calling handler <function generate_idempotent_uuid at 0x7eff1de77320>
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event before-call.cloudwatch-logs.PutLogEvents: calling handler <function inject_api_version_header_if_needed at 0x7eff1de7bdd0>
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/endpoint.py:101] Making request for OperationModel(name=PutLogEvents) with params: {'url_path': '/', 'query_string': '', 'method': 'POST', 'headers': {'X-Amz-Target': 'Logs_20140328.PutLogEvents', 'Content-Type': 'application/x-amz-json-1.1', 'User-Agent': 'Boto3/1.12.35 Python/3.7.6 Linux/4.15.0-1058-aws Botocore/1.15.35'}, 'body': b'{"logGroupName": "dev", "logStreamName": "django", "logEvents": [{"timestamp": 1585871099955, "message": "2020-04-02 23:44:59 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/asyncio/selector_events.py:58] Using selector: EpollSelector"}, {"timestamp": 1585871100212, "message": "2020-04-02 23:45:00 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/asyncio/selector_events.py:58] Using selector: EpollSelector"}]}', 'url': 'https://logs.us-west-2.amazonaws.com/', 'context': {'client_region': 'us-west-2', 'client_config': <botocore.config.Config object at 0x7eff1d108f10>, 'has_streaming_input': False, 'auth_type': None}}
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event request-created.cloudwatch-logs.PutLogEvents: calling handler <bound method RequestSigner.handler of <botocore.signers.RequestSigner object at 0x7eff1d108ed0>>
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py:210] Event choose-signer.cloudwatch-logs.PutLogEvents: calling handler <function set_operation_specific_signer at 0x7eff1de77200>
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:364] Calculating signature using v4 auth.
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:365] CanonicalRequest:
POST
/

content-type:application/x-amz-json-1.1
host:logs.us-west-2.amazonaws.com
x-amz-date:20200402T234504Z
x-amz-security-token:[snip]==
x-amz-target:Logs_20140328.PutLogEvents

content-type;host;x-amz-date;x-amz-security-token;x-amz-target
[snip]
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:367] StringToSign:
AWS4-HMAC-SHA256
20200402T234504Z
20200402/us-west-2/logs/aws4_request
[snip]
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/auth.py:369] Signature:
[snip]
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/endpoint.py:187] Sending http request: <AWSPreparedRequest stream_output=False, method=POST, url=https://logs.us-west-2.amazonaws.com/, headers={'X-Amz-Target': b'Logs_20140328.PutLogEvents', 'Content-Type': b'application/x-amz-json-1.1', 'User-Agent': b'Boto3/1.12.35 Python/3.7.6 Linux/4.15.0-1058-aws Botocore/1.15.35', 'X-Amz-Date': b'20200402T234504Z', 'X-Amz-Security-Token': b'[snip]==', 'Authorization': b'AWS4-HMAC-SHA256 Credential=[snip]/20200402/us-west-2/logs/aws4_request, SignedHeaders=content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=72d4315b645a1714dac464010cdb8de7561a77d4dc9f5f85fb7bd684d301349e', 'Content-Length': '433'}>
2020-04-02 23:45:04 DEBUG    [/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/urllib3/connectionpool.py:266] Resetting dropped connection: logs.us-west-2.amazonaws.com
```

Curiously, this file doesn't grow indefinitely if I wait a few minutes before interrupting the process.  Perhaps the file logger has successfully shut down, but the watchtower logger is still going?

(Edit: added formatting to log, with files and line numbers)

**Comment by georgedorn** [comment]
Setting `use_queues: False` in the logging config shows the original bug (#65) is back, with a recursion limit exceeded:

```(django) ubuntu@ip:/var/www/django$ ./manage.py shell
/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/watchtower/__init__.py:154: WatchtowerWarning: Failed to deliver logs: maximum recursion depth exceeded while calling a Python object
  warnings.warn("Failed to deliver logs: {}".format(e), WatchtowerWarning)
Fatal Python error: Cannot recover from stack overflow.

Thread 0x00007f7f97bde700 (most recent call first):
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/threading.py", line 296 in wait
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/threading.py", line 552 in wait
  File "/home/ubuntu/.local/lib/python3.7/site-packages/IPython/core/history.py", line 829 in run
  File "/home/ubuntu/.local/lib/python3.7/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "</home/ubuntu/.local/lib/python3.7/site-packages/decorator.py:decorator-gen-24>", line 2 in run
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/threading.py", line 890 in _bootstrap

Current thread 0x00007f7fbd508740 (most recent call first):
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py", line 403 in _alias_event_name
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py", line 355 in emit
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 670 in _emit_api_params
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 647 in _convert_to_request_dict
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 601 in _make_api_call
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 316 in _api_call
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/watchtower/__init__.py", line 145 in _submit_batch
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/watchtower/__init__.py", line 200 in emit
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 894 in handle
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 1586 in callHandlers
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 1524 in handle
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 1514 in _log
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 1366 in debug
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py", line 210 in _emit
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py", line 228 in emit
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/hooks.py", line 356 in emit
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 679 in _emit_api_params
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 647 in _convert_to_request_dict
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 601 in _make_api_call
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/botocore/client.py", line 316 in _api_call
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/watchtower/__init__.py", line 145 in _submit_batch
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/site-packages/watchtower/__init__.py", line 200 in emit
  File "/home/ubuntu/miniconda/envs/django/lib/python3.7/logging/__init__.py", line 894 in handle
[repeat until recursion limit]
  ...
Aborted (core dumped)
(django) ubuntu@ip-10-0-2-88:/var/www/django$ 
```

**Comment by kislyuk** [comment]
Thank you for your interest in Watchtower. Nothing about this behavior is strictly a bug: watchtower uses botocore to send log messages to AWS. Botocore and its dependency urllib3 emit logs with every request when its logger's log level is set to DEBUG. Setting the root log level to DEBUG without also overriding the botocore log level will cause an infinite loop since the log handler won't quit until all messages are delivered, but each delivery will generate more messages.

I recognize that this isn't the expected behavior, so I implemented a workaround in v1.0.2. The log level for botocore will be forced to INFO or higher while watchtower is shutting down. This should help avoid the issue that you encountered. Let me know if you still see it happen with this release.
